### PR TITLE
Ghost Cafe Additions

### PR DESCRIPTION
**Pull request by ForrestWick**
## About The Pull Request
The ghost cafe has been upgraded a bit, and now features a fun little combat arena with an entire armory prep room suited up with all sorts of weapons and armor to play with (all pre-existing items, nothing new added). In addition: a construction zone is sitting on the eastern end of the cafe, with the entrance having crates with more than enough materials and tools for you to build whatever vibe huts and fuck shacks you want.

In addition, the ghost cafe itself now has a few designated areas made for it specifically. The areas are there to keep atmos in check for the most part, as well as provide automatic pacifism when you enter or exit certain portions of the cafe. This way people can't take the guns out of the combat zone and start killing people that just wanna chill. As an added precaution: the combat zone has turnstiles that you can't shoot through anyway.

The old melee rage cage has been replaced with a dodgeball court, but it runs on the same principles as the combat zone. Pacifism is disabled there, so if you wanna do something more destructive than dodgeball in a small, open space, there you go.

Many thanks to Putnam for helping and holding my hand through the pacifism area coding. They're a damn saint.

## Why It's Good For The Game
The old cafe was a good addition, but doesn't provide a lot for people that wound up dying early in a round and want to mess around with stuff. This PR aims to give these people a nice, isolated zone where they can fight in the kill zone or build in the construction zone if they don't feel like ERPing or doing botany.

## Changelog
:cl:
add: Added two new zones and three new areas for the ghost cafe, with the areas having automatic pacifism when you enter and still retaining the "dust on leaving" effect if you escape the cafe itself.
tweak: replaced the melee cage in the cafe with a dodgeball court
/:cl:

### CODE REVIEW

**Comment by MrJWhit** [comment]
I'm completely against this, I don't think this is good for the server. It promotes people from actively playing the game with everyone else, and doesn't (by design) feed into the actual game.

**Comment by lolman360** [comment]
more stuff for ghosts is good (see mafia)

**Comment by ForrestWick** [comment]
> I'm completely against this, I don't think this is good for the server. It promotes people from actively playing the game with everyone else, and doesn't (by design) feed into the actual game.

This has been argued repeatedly in the Discord server. Nothing is going to stop people from going to the ghost cafe if they really wanted to. At worst this will be a few days of people going to the cafe to mess with the new toys, but then after that it'll be people who didn't wanna play the round anyway or got ganked in the first 20 minutes/cryo'd because of the game mode.

**Comment by MrJWhit** [comment]
I also don't think that the ghost cafe is healthy for the game in the first place, a sectioned off area where players are babied is actively unhelpful for learning the game, and promoting learning.

If someone leaves the game, they can do any of the, many, other ghostroles that _do_ feedback into the game itself. Learn botany in lavaland. Learn science as a golem. Learn combat and engineering in oldstation. Learn how to fight people as an ashwalker. This actively pulls people away from doing the basic gameplay loop by turning it into a damn chatroom.

**Comment by ForrestWick** [comment]
You forget that Goon has a ghost cafe.

**Comment by MrJWhit** [comment]
Ok. Cool. We're not Goon. We're not Skyrat. Porting things over because other people have it, is not an argument.

**Comment by ForrestWick** [comment]
It is when Goon had it and they turned out just fine despite having it. Don't forget that Putnam made the cafe for Cit initially, and then Owai made improvements.

**Comment by MrJWhit** [comment]
We are not Goon, we have a different playerbase than Goon. We have a different server with different rules, roleplaying requirements, a completely different codebase, and so on. It's not an argument on why it should be brought over by itself. I can't see a reason on how this improve gameplay for the server.

**Comment by MrJWhit** [comment]
Besides, this is in active development over in skyrat, there are still fixes being applied to it. There will be fixes applied to those fixes as well. At the very least wait a week to confirm that there isn't any unforeseen consequences before bringing it over.

**Comment by ForrestWick** [comment]
This isn't a direct port of Skyrat's version. I did the map edits for this one from scratch and Putnam pretty much did the pacifism area code for me.

**Comment by Putnam3145** [comment]
Mmh, I kinda really don't want the ghost cafe to be too... big. But this depends on things. I might testmerge this and sic @timothyteakettle on it to see if it's more breakable than the previous, and so on.

**Comment by MrJWhit** [comment]
I can't find a way to open this in either strongdmm, or byonds built in map editor, and mapdiffbot doesn't appear to be running, I'm not fully convinced that it's functional right now.

**Comment by ForrestWick** [comment]
> I can't find a way to open this in either strongdmm, or byonds built in map editor, and mapdiffbot doesn't appear to be running, I'm not fully convinced that it's functional right now.

I can assure you I ran as much testing as I could and it worked in local.

**Comment by timothyteakettle** [comment]
i like giving ghosts something to do
testmerge this and then just have people request changes as needed

**Comment by MrJWhit** [comment]
I meant for any adjustments that are needed preemptively.

File: code/game/objects/structures/ghost_role_spawners.dm [hunk 1]
```diff
@@ -670,11 +670,10 @@
 		SSquirks.AssignQuirks(new_spawn, new_spawn.client, TRUE, TRUE, null, FALSE, new_spawn)
 		new_spawn.AddElement(/datum/element/ghost_role_eligibility, free_ghosting = TRUE)
 		new_spawn.AddElement(/datum/element/dusts_on_catatonia)
-		new_spawn.AddElement(/datum/element/dusts_on_leaving_area,list(A.type,/area/hilbertshotel))
+		new_spawn.AddElement(/datum/element/dusts_on_leaving_area,list(A.type,/area/hilbertshotel | /area/centcom/holding/cafe | /area/centcom/holding/cafebuild | /area/centcom/holding/cafewar))
```
Comment (Putnam3145) [inline code/game/objects/structures/ghost_role_spawners.dm]: these should just be commas, not |s, the | is for putting two lists together while this is all in one list

Comment (Putnam3145) [reply code/game/objects/structures/ghost_role_spawners.dm]: i should note that this is runtiming, which breaks everything below this line

Comment (Putnam3145) [reply code/game/objects/structures/ghost_role_spawners.dm]:
```suggestion
		new_spawn.AddElement(/datum/element/dusts_on_leaving_area,list(A.type,/area/hilbertshotel, /area/centcom/holding/cafe, /area/centcom/holding/cafebuild, /area/centcom/holding/cafewar))
```

**Comment by Tupinambis** [comment]
I love the idea, but there are quite a few kinks to hammer out. The first is exploits which I've already pinged you about, but another big thing would be making a SAFE staging area both for construction equipment and the arena. Some EVA ready suits for construction would also be nice for bonus creativity.

**Comment by Echopsychotic** [comment]
As someone who does play skyrat, and sometimes citadel I use the ghost cafe quite often. I think this is a perfect use for a learning curve for players if they do not want to do the current round and be interrupted by the current antags

**Comment by Detective-Google** [comment]
yeah and how much of this stuff is already present in VR? w-what? almost all of it?

**Comment by MrJWhit** [comment]
Or you can learn in-game with other people helping you.

**Comment by ForrestWick** [comment]
Update:

The cafe war zone and prep area uses turnstiles with increased health and damage resistance, along with having invisible, unbreakable atmos fans. The RCDs that were added are downgraded and can no longer deconstruct, but will still be able to build stuff. The dodgeball court no longer has breakable walls.

**Comment by ForrestWick** [comment]
Need to redo this from scratch, issues arose and previous branch is gone

**Pull request closed without merging**